# **-01-Basics-of-Python-Programming © by A.Richter**

<img src=88x31.png  width="10%">

**Preface:** "...A programming language gives you the power to express some ideas, while 
limiting your abilities to express others.“  [LARRY CUBA, computer-animation artis](https://en.wikipedia.org/wiki/Larry_Cuba)

This statement independent of a programming language points to the expression of ideas and creativity. The computer today is used as a tool in the hands of engineers, scientists or any other creative person. Just like a craftsman uses a screwdriver todays knowledgeworker use a specific programming language. But as we all know the toolbox of a craftsman accomodates much more tools than only screwdrivers.

These notebooks are collections of usefull tools including explanations of theory ranging from computerscience to earthsciences, mathmatics and physics. The start is an introduction to the python language. Every notebook has a preface describing roughly the content and the pupose of the discussed topics.

# Overview of the notebook:
    
   1. [Basic Libray Imports, variables and printing](#basiclibraryimports)
   2. [The Zen of Python](#zen)
   3. [Intro to bits and bytes](#bits)
   4. [A little bit of math in Python](#math)
   5. [Python datatypes](#datatypes)
   6. [If-Else Decision Constructions](#ifthen)
   7. [Loop Constructions](#loops)
   8. [Folder and File operations](#folders)
   9. [Functions](#functions)
   10. [Basics of nummeric python (numpy) and scientific python (scipy)](#numpybasics)
   11. [Revision of statistical Tools](#stats)
   12. [Preprocessing in Data Analysis](#preprocessing)

**Sources: **
www.introtopython.org  and    http://www.scipy-lectures.org/   


But before we start with learning Python

you should know at least a bit how to use the Commandline **CMD** for WINDOWS or **Terminal** for LINUX & MAC

https://www.thomas-krenn.com/de/wiki/Cmd-Befehle_unter_Windows


1) How do start the Commandline/Terminal in Windows, Linux, macOS? 

2) How do you change your position in the filesystem to a specific folder?

3) How do you create a new txt file in that folder?

4) How do you start Jupyter Notebook in this folder? 


===================================================================

## 1. Basic Libray Imports, variables and printing <a name="basiclibraryimports" ></a>

In general Python is build modular, which means for special things look for the right module and simply import it...

Basic Python Libraries or modules can be imported like this:

In [1]:
import sys

The keyword **'import'** followed by the name of the library you want to use.

There are some specials hidden in the python libraries, for example here:

https://wiki.python.org/moin/UsefulModules                           # Overview

https://www.glowscript.org/#/user/GlowScriptDemos/folder/Examples/   # 3D Examples

https://pypi.org/project/geopy/                                      # Geo tools 

When is a special module needed???

For example see the equation  $10 \cdot 10$ = $10^2$ = 100

That implies that the logarithm to base 10 of 100 must be equal to 2.

So how can I calculate this in python?


In [3]:
log10(100)

NameError: name 'log10' is not defined

this error message means the base python module library does not know the function log10()

But the python module **math** has a definition of log10() and is therefor needed to calculate log10(100).

So at first I have to import the **math** module

In [19]:
import math

and than I can use it in this way:

In [20]:
math.log10(100)

2.0

A modular function is always called in the form: "**moduleName.functionName(parameters)**" 

### Variables in Python

In [1]:
x=3

Python is very humble concerning variables. There is no need for prior definitions or declaration of the datatype (like in C/C++). Variable are simply assigned to values and can be used immediately.

**The Command "whos" is displaying the current use of variables in the main memory(RAM)**


In [2]:
whos

Variable   Type    Data/Info
----------------------------
x          int     3


### Print out something in python

Printing of an expression or a variable is easily done in this way:

In [13]:
print("HELLO WORLD!")
print("-------------")
print("+-/* Sum: {} ".format(100))

HELLO WORLD!
-------------
+-/* Sum: 100 


Keep in mind that the older versions of python (e.g. Python 2.7) was not using the brackets. Python 3 is very strict in this case and forcing the use of **print("Text2print", variabes)** 

In [4]:
print("The number assigned to the Variable x is",x)

The number assigned to the Variable x is 3


## The Zen of Python <a name="zen" ></a>


In [5]:
import this
#import codecs
#print(this.s.decode('rot13'))
print(this.s)
print(70*'=')

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!
Gur Mra bs Clguba, ol Gvz Crgref

Ornhgvshy vf orggre guna htyl.
Rkcyvpvg vf orggre guna vzcyvpvg.
Fvzcyr vf orggre guna pbzcyrk.
Pbzcyrk vf or

In [10]:
d = {}
for c in (65, 97):
    for i in range(26):
        d[chr(i+c)] = chr((i+13) % 26 + c)

print("".join([d.get(c, c) for c in this.s]))

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


## Intro to Bits and Bytes <a name="bits" ></a>


Computer Science is based on the idea of binary states, whether it's **Light: on/off** or **electr. Voltage: high/low** 


BIT = binary digit – englisch | in german "binäre Ziffer" or  "Binärziffer"

one bit can have two conditions $0$ or $1$ and can be expressed in a vector like this

1bit = $\begin{bmatrix}
0          \\
1        \end{bmatrix}$ 


On a computer numbers can be used in form of the binary system or in form of the decimal system.

In [5]:
print('binary equal decimal')
print('110001   =   ', 1*(2**5)+1*(2**4)+0*(2**3)+0*(2**2)+0*(2**1)+1*(2**0))

binary equal decimal
110001   =    49


But how to calculate this:

In [7]:
x=1*(2**5)+1*(2**4)+0*(2**3)+0*(2**2)+0*(2**1)+1*(2**0)
print(x)

49


or

In [8]:
print(int('110001', 2))

49


1 **Byte** = 8 Bits
= $\begin{bmatrix}
0  & 0&0&0&0&0&0&0       \\
1  & 1&1&1&1&1&1&1     \end{bmatrix}$

so in general 8 options to decide for 1 or 0


In [2]:
print("10010101")
print(int('10010101', 2))

10010101
149


This means that the binary number: 10010101 represents the decimal number: 49

**Gigabytes vs. Gigibytes**

There are different aproaches of defining bigger bunches of information.
For example are **1000 Bytes** called a **kilobyte** which is based on the well known **decimal scale**.

$1000^2$ Bytes = 1MB

$1000^3$ Bytes = 1GB


In [11]:
print(1000**3, "Bytes = 1 GB = gigabyte")

1000000000 Bytes = 1 GB = gigabyte


On the other hand using a binary scale with

1024**3  (GiB)  gibibyte

https://en.wikipedia.org/wiki/Gigabyte

try it yourself...

In [1]:
print(1024**3, "Bytes = 1 GiB = Gigibyte")

1073741824 Bytes = 1 GiB = Gigibyte


## A little bit of math in Python <a name="math" ></a>

Check the Value of a Variable - compare it to another one

In [15]:
x=3

This means that the memory assigned to the variable x is filled with the value 3

In [16]:
x==5

False

The double equal sign (==) is a completly different operation. It implies the comparison of the value stored in the variable x with the value 5. So in other words is 3 equal to 5? No!

In [17]:
x==3

True

But 3 is equal to 3. That's right!

The basic mathmatical operations are expressed for example like this:

In [34]:
x=3
y=5
print(x+y)
print(x-y)
print(x*y)
print(x/y)

8
-2
15
0.6


Division of values in Python 3 is different compared to Python 2 --> Python-2 is using "Int" Python-3 always converts to "Float" before any calculation is done!!!

In [13]:
5/3

1.6666666666666667

In [2]:
5.0/3.0

1.6666666666666667

In [3]:
5**3

125

calculates the the power: $5^3 =5 * 5 * 5=125$


## Python Datatypes <a name="datatypes" ></a>

We will now have a look at basic Datentypes like integer ("**int**") and floating point numbers like "**float**" or "**double**" and of course character types like "**string**".

### Typ: Sting

In [5]:
import sys
S="D:\BHT-BERLIN\nDaten\datei1.shp"
print(S)
print(type(S))
print("Size: {} Bytes".format(sys.getsizeof(S)))

D:\BHT-BERLIN
Daten\datei1.shp
<class 'str'>
Size: 79 Bytes


oh look what happened here? The String S was cut in the middle...but why???

the folder  **"\nDaten"**  has a problem, the **"\n"** is interpreteed as a line break!!! 

try ist yourself with a different folder path of your choice...

...better use it this way: "/" instead of "\" to separate folders

In [4]:
S="D:/BHT-BERLIN/nDaten/datei1.shp"
print(S)
print(type(S))
print("Size: {} Bytes".format(sys.getsizeof(S)))


D:/BHT-BERLIN/nDaten/datei1.shp
<class 'str'>
Size: 80 Bytes


or use a so called raw string format **r"... \ ... \ ..."**

In [6]:
S=r"D:\BHT-BERLIN\nDaten\datei1.shp"
print(S)
print(type(S))
print("Size: {} Bytes".format(sys.getsizeof(S)))

D:\BHT-BERLIN\nDaten\datei1.shp
<class 'str'>
Size: 80 Bytes


which will work in the same way

In [22]:
y="BHT"
z="Berlin"
print(y+" - "+z)
print(15*"-")
print(y) 
print("Size: {} Bytes".format(sys.getsizeof(y)))
print(z) 
print("Size: {} Bytes".format(sys.getsizeof(z)))

BHT - Berlin
---------------
BHT
Size: 52 Bytes
Berlin
Size: 55 Bytes


**Tye: Int**

In [27]:
I=0
print(type(I))
print("Size:{} Bytes".format(sys.getsizeof(I)) )

<class 'int'>
Size:24 Bytes


In [31]:
I=1
print(type(I))
print("Size:{} Bytes".format(sys.getsizeof(I)) )

<class 'int'>
Size:28 Bytes


In [30]:
print("Size:{} Bytes".format(sys.getsizeof(2 ** 63)) )

Size:36 Bytes


**Tye: Float**

In [31]:
F=2.0
#x=6.45678987665454433345666
print(type(F))
print("Size:{} Bytes".format(sys.getsizeof(F)))

<class 'float'>
Size:24 Bytes


In [30]:
F=6.45678987665454433345666
print(type(F))
print("Size:{} Bytes".format(sys.getsizeof(F)))

<class 'float'>
Size:24 Bytes


==> Why are the sizes of INT and FLOAT different?

https://stackoverflow.com/questions/10365624/sys-getsizeofint-returns-an-unreasonably-large-value

https://stackoverflow.com/questions/29826523/python-float-precision-float

## If-Else Decision Constructions <a name="ifthen" ></a>

In Python this is used in this way: 

In [9]:
x=5

if x<0:
    print(x)
    print(2*x)
else:
    print(100-x)

95


Keep an eye on the block that is executed when the if condition (here $x<0$) was evaluated successfully. All statements have to start in the same column (greater than column 1) otherwise python will rise an error.

In [6]:
x=-5

if x<0:
    print(x)
     print(2*x)
else:
    print(100-x)

IndentationError: unexpected indent (<ipython-input-6-e92e48d70095>, line 5)

In [8]:
x=-5

if x<0:
    print(x)
    print(2*x)
else:
    print(100-x)

-5
-10


this is also mandatory for the **else** block but obviously we have here only on line of code.

Other programming languages like C or Java require brackets to start or to end a block construktion. Python just interpretes all statements in the same column to be part of the current block. 

Regarding the condition that is evaluated in the first if-statement all sorts of logic examples like **...and...**,  **...or...**, **not...**, **...greater than...** constructions can be used:


   - Equals: a == b
   - Not Equals: a != b
   - Less than: a < b
   - Less than or equal to: a <= b
   - Greater than: a > b
   - Greater than or equal to: a >= b
   - ...

Different cases can also be expressed with the **elif** branch: 

In [11]:
a = 33
b = 33
if b > a:
    print("b is greater than a")
elif a == b:
    print("a and b are equal")

a and b are equal


## Loop Constructions <a name="loops" ></a>

Loops are the second basic pillar used in programming languages next to conditionals.

There are different loops available in python, but the most widely used loop is the **for-loop** 

### while-Loop

This is a simple example for a while-loop construction:

In [11]:
x=0
while x<10:
    print(x)
    x=x+1

0
1
2
3
4
5
6
7
8
9


### for-loop

The power of the for-loop construction is mainly that it is implemented as a **for-each algorithm**.
This means that the user does not need to know how many item there are to be iterarted over before the loop starts. The for loop works on python lists that inherrently know their start, end and their number of items in between.

In [12]:
for i in range(35,40):
    print(i)

35
36
37
38
39


In [37]:
for i in range(35,55,5):
    print(i)

35
40
45
50


The function **range(START,STOP,STEP)** provides a highly flexible tool in combination with the for-Loop

Be aware that the specified STOP Value is always evaluated at the index before the value itself.


## Folder and File Operations <a name="folders" ></a>

To Search a Folder for a specific File Type, the following construction can be used:

In [44]:
# list a Folder with Tif + Tfw Files

import glob
Folder='figures/'
#Folder='../../_Data_ori_Tif/AOI2_small_6_Test/'
# https://docs.python.org/2/library/glob.html 
# http://stackoverflow.com/questions/3207219/how-to-list-all-files-of-a-directory-in-python

JPGs=glob.glob(Folder+"*.jpg")
PNGs=glob.glob(Folder+"*.png")

for j in JPGs:
    print(j)

print(30*'-') 
    
for p in PNGs:
    print(p)

figures/iris_versicolor.jpg
figures/std.jpg
figures/doug.jpg
figures/petal_sepal.jpg
figures/iris_virginica.jpg
figures/iris_setosa.jpg
------------------------------
figures/Line-2.png
figures/Line-3.png
figures/Line-1.png
figures/feynman.png
figures/blunder.png
figures/supervised_scikit_learn.png
figures/ml-eng.png
figures/Quad-Rot-1.png
figures/FE_Class.png
figures/Eigenvalues-Eigenvectors.png
figures/Bootstrapp_Example2.png


## Functions <a name="functions" ></a>

Functions are a great and important topic in Python. We have already used some of them in the examples before.

Simply write the litteral "def" at the beginning of a line
followed by the name of the function "MyFunct()" 

In [37]:
def MyFunct(af,bf):
    cf=af+bf
    return cf

a=4
b=4
print("a =",a,"| b =",b)
print(2*"-------")

print("Summe(a,b): {}".format(MyFunct(a,b)))

a = 4 | b = 4
--------------
Summe(a,b): 8


In [40]:
# Functions: indent blocks with 4 spaces
def hello_world(name, times=3):
    print("Say",name*times)
    return 

hello_world("yes")

Say yesyesyes


A function to compute the exponential function  $f(x)=e^x$ by a taylor approximation:

$$e^x= \sum_{n=0}^{\infty} \frac{x^n}{n!} =1+x+\frac{x^2}{2!}+\frac{x^3}{3!}+\frac{x^4}{4!}+...$$

In [21]:
import math
def taylorexp(inp, acc):
    sum=0
    for i in range(0,acc):
        sum=sum+(inp**i)/math.factorial(i)
    return sum

In [35]:
math.exp(4)

54.598150033144236

In [36]:
print(taylorexp(4,100))
print(taylorexp(4,100) - math.exp(4))

54.598150033144265
2.842170943040401e-14


## Basics of Nummeric Python (numpy) and Scientific Python (scipy) <a name="numpybasics" ></a>


Many scientific problems can be described with algebraic expressions in form of a tensor formalism. 

We need to distinguish between the used dimensions of arrays



Numpy matrices are strictly 2-dimensional, while numpy arrays (ndarrays) are N-dimensional. Matrix objects are a subclass of ndarray, so they inherit all the attributes and methods of ndarrays.

The main advantage of numpy matrices is that they provide a convenient notation for matrix multiplication: if a and b are matrices, then a*b is their matrix product.

In [41]:
import numpy as np

a=np.mat('4 3; 2 1')
b=np.mat('1 2; 3 4')
print(a)
# [[4 3]
#  [2 1]]
print(b)
# [[1 2]
#  [3 4]]
print(a*b)
# [[13 20]
#  [ 5  8]]

[[4 3]
 [2 1]]
[[1 2]
 [3 4]]
[[13 20]
 [ 5  8]]


In [42]:
a=np.array([[4, 3], [2, 1]])
b=np.array([[1, 2], [3, 4]])
print(a@b)

[[13 20]
 [ 5  8]]


In [7]:
print(np.zeros((5,5)))

[[ 0.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.]]


In [6]:
print(np.ones((3,3)))

[[ 1.  1.  1.]
 [ 1.  1.  1.]
 [ 1.  1.  1.]]


A simple 4x4 Matrix or an 4x4 Array in nummeric Python is initilized like this:

In [21]:
A=np.array([[2,4,3,7],[1,2,3,4],[-2,4,5,7],[4,3,7,9]])
print(A)

[[ 2  4  3  7]
 [ 1  2  3  4]
 [-2  4  5  7]
 [ 4  3  7  9]]


Calculate the sum of each column

In [22]:
b=np.sum(A, axis=0)
print(b)

[ 5 13 18 27]


Calculate the sum of each row

In [23]:
c=np.sum(A, axis=1)
print(c)

[16 10 14 23]


Pick an element from the Array by its index

In [24]:
d=A[2,1]
print(A)
print("  ") 
print('The Value at index [row=2,col=1] is: {} '.format(d))

print('Size of the Value (.itemsize): {} Bytes '.format(d.itemsize))
print('Size of the Value (.nbytes): {} Bytes '.format(d.nbytes))

[[ 2  4  3  7]
 [ 1  2  3  4]
 [-2  4  5  7]
 [ 4  3  7  9]]
  
The Value at index [row=2,col=1] is: 4 
Size of the Value (.itemsize): 8 Bytes 
Size of the Value (.nbytes): 8 Bytes 


Pick a row from the array and play with the index in different ways

In [26]:
print('Row one to end: all Rows')
print('-------------')
print(A[1,:])
print(' ')
print('All Rows with specific Index')
print('-------------')
print(A[1,0:4])
print(' ')
print('All without the 1st Row ')
print('-------------')
print(A[1,1:4])
print('-------------')

Row one to end: all Rows
-------------
[1 2 3 4]
 
All Rows with specific Index
-------------
[1 2 3 4]
 
All without the 1st Row 
-------------
[2 3 4]
-------------


Change the Array, **replace the second row**

In [27]:
print(A) 
A2=A
print(2*'---------')
A2[1,:]=[5,6,7,8]

print(A2)

[[ 2  4  3  7]
 [ 1  2  3  4]
 [-2  4  5  7]
 [ 4  3  7  9]]
------------------
[[ 2  4  3  7]
 [ 5  6  7  8]
 [-2  4  5  7]
 [ 4  3  7  9]]


Creating a **Vector** in diffenrent ways

In [29]:
e=np.arange(0,11)
print(e)

[ 0  1  2  3  4  5  6  7  8  9 10]


In [32]:
f=np.arange(1,8.5,0.5)
print(f)

[ 1.   1.5  2.   2.5  3.   3.5  4.   4.5  5.   5.5  6.   6.5  7.   7.5  8. ]


Matrix multiplication

In [39]:
print(A," \t\n\n",A2) 

[[ 2  4  3  7]
 [ 5  6  7  8]
 [-2  4  5  7]
 [ 4  3  7  9]]  	

 [[ 2  4  3  7]
 [ 5  6  7  8]
 [-2  4  5  7]
 [ 4  3  7  9]]


In [33]:
np.dot(A,A2)

array([[ 46,  65,  98, 130],
       [ 58, 108, 148, 204],
       [ 34,  57,  96, 116],
       [ 45,  89, 131, 182]])

Element-wise multiplication

In [34]:
A*A2

array([[ 4, 16,  9, 49],
       [25, 36, 49, 64],
       [ 4, 16, 25, 49],
       [16,  9, 49, 81]])

References: 

https://stackoverflow.com/questions/4151128/what-are-the-differences-between-numpy-arrays-and-matrices-which-one-should-i-u

## Revision of statistical-Tools <a name="stats" ></a>

**Mean Value and Standard-deviation**

Generate a list of 20 random values,
around 50

In [103]:
x=50+10*np.random.randn(20,1)
print(x)

[[ 41.59482751]
 [ 59.43655008]
 [ 33.42469536]
 [ 40.5579171 ]
 [ 71.16557896]
 [ 27.10040817]
 [ 60.04444181]
 [ 58.30891218]
 [ 38.6730831 ]
 [ 52.00083727]
 [ 39.15168817]
 [ 31.5525942 ]
 [ 52.49034837]
 [ 36.63040418]
 [ 55.74190451]
 [ 55.30373652]
 [ 52.37174155]
 [ 57.31837229]
 [ 37.67475312]
 [ 50.15780303]]


**Calculte the Mean of the Values:**

$ \bar{X} =\frac{\sum\nolimits_{i=1}^{20} x_i}{20}$ the long and old-school way

In [104]:
Xmean=np.sum(x)/20

print(Xmean)

47.5350298735


or calculate it with a numpy function:

In [105]:
np.mean(x)

47.535029873541738

try it  yourself:

**Calculate the Standard Deviation of the Values:**

$ std = \sqrt{\frac{\sum\nolimits_{i=1}^{20}(x_i - \bar{X})^2}{20-1}}$

In [106]:
np.sqrt( np.sum((x-np.mean(x))**2)/19 ) 

11.679563299046686

In [107]:
np.std(x)

11.383830151298655

but how is numpy calculating here? Be carefull is you use the function and read the documentation first...!!! 

In [108]:
print(np.std(x,ddof=1) )

11.679563299


https://stackoverflow.com/questions/34050491/standard-deviation-in-numpy

https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.std.html

ddof : int, optional

ddof means **D**elta **D**egrees **o**f **F**reedom. The divisor used in calculations is N - ddof, where N represents the number of elements. **By default ddof is zero.**


In [65]:
numbers=np.array([[6,7,8,9,10]])
print(numbers)

[[ 6  7  8  9 10]]


In [66]:
np.mean(numbers)

8.0

In [94]:
np.std(numbers,ddof=1)

1.5811388300841898

is called the proper calculated std, with n-1 degrees of freedom

In [95]:
np.std(numbers)

1.4142135623730951

is called a Maximum Likelihood estimation of the std

explaind with a very simple example, see source...

<img src="figures/std.jpg">

source: https://www.frustfrei-lernen.de/mathematik/standardabweichung-berechnen.html

In the following notebook **-01a-Optimizing-your-SourceCode** I will show some usefull construction and techniques that ensure a better performance than the classical programming paradigm.


## Preprocessing in Data Analysis <a name="preprocessing" ></a>

Show Data in its original content - just view

In [ ]:
#Anzeigen der Bilder und Histogramme

from scripts.BasisFunctions import showImage

    
    
for layer in sorted(LayerTrain):
    print layer
    showImage( LayerTrain.get(layer))
    #print Layers.get(layer)

- get number and index of noData Points (NaN)
print('Anzahl NaN in Dataset-Name: {}'.format(np.sum(np.isnan(../Path-to-Dataset)*1)))


- get number and index of (infinity positions)
print('...these positions are set to Zero(0) :')
Dataset[np.isinf(Dataset)]=0


:Mean od Median insection

In [ ]:
#Prüfung auf Datenvollständigkeit
#NaN - Not a Number
# Infty....

print 'Anzahl infty in TIF_EucDist_faults: {}'.format(np.sum(np.isinf(TIF_r_EucDist_faults)*1))

TIF_r_EucDist_roads[np.isinf(TIF_r_EucDist_roads)]=0
TIF_r_EucDist_takhar[np.isinf(TIF_r_EucDist_takhar)]=0
TIF_r_DistStreams[np.isinf(TIF_r_DistStreams)]=0

print 'Anzahl infty in TIF_EucDist_faults: {}'.format(np.sum(np.isinf(TIF_r_EucDist_faults)*1))
print 'Anzahl infty in TIF_EucDist_roads: {}'.format(np.sum(np.isinf(TIF_r_EucDist_roads)*1))
print 'Anzahl infty in TIF_EucDist_takhar: {}'.format(np.sum(np.isinf(TIF_r_EucDist_takhar)*1))
print 'Anzahl infty in TIF_DistStreams: {}'.format(np.sum(np.isinf(TIF_r_DistStreams)*1))


Change the scaling of the values in case they are unballanced.


In [ ]:

print('for example apply a logarithmic scale to Distance Rasters')

TIF_r_EucDist_faults =np.log(TIF_r_EucDist_faults)


Change Datatype of Rasters in case they do not match to the rest of the data

In [ ]:

TIF_r_precip=np.float32(skio.imread(Folder+'TIF_precip.tif'))
print 'TIF_r_precip Shape: ', TIF_r_precip.shape,TIF_r_precip.dtype
Layers['TIF_r_precip']=TIF_r_precip


Create a Python dictionary to work efficintly with a bunch of raster arrays

In [ ]:
Layers=dict()
Layers['TIF_r_dem'] = TIF_r_dem
Layers['TIF_r_precip']=TIF_r_precip

Generate/Split off new layers from specific numbers of a single array or reorganize the layers

In [ ]:
Layers['TIF_n_LanduseRAL']=((Layers.get('TIF_n_lc93_32') ==1) | (Layers.get('TIF_n_lc93_32') ==5))*1.0
print 'TIF_n_LanduseRAL ', Layers['TIF_n_LanduseRAL'].shape
        
Layers['TIF_n_LanduseRL']=(Layers.get('TIF_n_lc93_32') ==2)*1.0
print 'TIF_n_LanduseRL ', Layers['TIF_n_LanduseRL'].shape
        
Layers['TIF_n_LanduseIRA']=((Layers.get('TIF_n_lc93_32') ==4) | (Layers.get('TIF_n_lc93_32') ==9) |(Layers.get('TIF_n_lc93_32') ==10))*1.0
print 'TIF_n_LanduseIRA' ,Layers['TIF_n_LanduseIRA'].shape


In [ ]:
Layers['TIF_n_VOLC']= (Layers.get('TIF_n_lith_type') == 1)*1.0  # % Volcanic ...... Vulkanisches Gestein
print 'TIF_n_VOLC ', Layers['TIF_n_VOLC'].shape
        
Layers['TIF_n_SEDI']= (Layers.get('TIF_n_lith_type') == 2)*1.0  # % Sedimentary ... Sedimente
print 'TIF_n_SEDI ', Layers['TIF_n_SEDI'].shape
       
Layers['TIF_n_PLUT']= (Layers.get('TIF_n_lith_type') == 3)*1.0  # % Plutonic....... Plutonisches Gestein
print 'TIF_n_PLUT ', Layers['TIF_n_PLUT'].shape


Reshaping Arrays and Vectors 